src/iceberg/update/update_partition_spec.cc

```diff
@@ -0,0 +1,488 @@
+  std::string_view source_name = "unknown";
+  if (field_result.has_value() && field_result.value().has_value()) {
+    source_name = field_result.value().value().get().name();
+  }
+
+  return transform->GeneratePartitionName(std::string(source_name));
+}
+
+bool UpdatePartitionSpec::IsTimeTransform(const std::shared_ptr<Transform>& transform) {
+  switch (transform->transform_type()) {
+    case TransformType::kYear:
+    case TransformType::kMonth:
+    case TransformType::kDay:
+    case TransformType::kHour:
+      return true;
+    default:
+      return false;
+  }
+}
+
+bool UpdatePartitionSpec::IsVoidTransform(const PartitionField& field) {
+  return field.transform()->transform_type() == TransformType::kVoid;
+}
+
+void UpdatePartitionSpec::CheckForRedundantAddedPartitions(const PartitionField& field) {
+  if (IsTimeTransform(field.transform())) {
+    if (added_time_fields_.contains(field.source_id())) {
+      AddError(ErrorKind::kInvalidArgument,
+               "Cannot add redundant partition field: {} conflicts with {}",
+               field.ToString(), added_time_fields_.at(field.source_id()));
+      return;
+    }
+    added_time_fields_.emplace(field.source_id(), field.ToString());
+  }
+}
+
+std::unordered_map<std::string, const PartitionField*>
+UpdatePartitionSpec::IndexSpecByName(const PartitionSpec& spec) {
+  std::unordered_map<std::string, const PartitionField*> index;
+  for (const auto& field : spec.fields()) {
+    index.emplace(std::string(field.name()), &field);
+  }
+  return index;
+}
+
+std::unordered_map<UpdatePartitionSpec::TransformKey, const PartitionField*,
+                   UpdatePartitionSpec::TransformKeyHash>
+UpdatePartitionSpec::IndexSpecByTransform(const PartitionSpec& spec) {
+  std::unordered_map<TransformKey, const PartitionField*, TransformKeyHash> index;
+  index.reserve(spec.fields().size());
+  for (const auto& field : spec.fields()) {
+    TransformKey key{field.source_id(), field.transform()->ToString()};
+    index.emplace(key, &field);
+  }
+  return index;
+}
+
+void UpdatePartitionSpec::BuildHistoricalFieldsIndex() {
+  const TableMetadata* base_metadata = transaction_->base();
+  if (base_metadata == nullptr) {
+    return;
+  }
+
+  // Count total fields across all specs to reserve capacity
+  size_t total_fields = 0;
+  for (const auto& partition_spec : base_metadata->partition_specs) {
+    total_fields += partition_spec->fields().size();
+  }
+  historical_fields_.reserve(total_fields);
+
+  // Index all fields from all historical partition specs
+  // Later specs override earlier ones for the same (source_id, transform) key
+  for (const auto& partition_spec : base_metadata->partition_specs) {
+    for (const auto& field : partition_spec->fields()) {
+      TransformKey key{field.source_id(), field.transform()->ToString()};
+      // Use emplace to only insert if key doesn't exist, preserving first occurrence
+      // This ensures we get the earliest field ID for recycling
```

Comment:
Is this correct? IIUC, `base_metadata->partition_specs` does not guarantee any order of the partition specs.